crates/core/src/metadata_store.rs

```diff
@@ -299,28 +306,43 @@ impl MetadataStoreClient {
         &self,
         key: ByteString,
     ) -> Result<Option<T>, ReadError> {
-        let value = self.inner.get(key).await?;
+        let start_time = Instant::now();
+        counter!(METADATA_CLIENT_GET_TOTAL, "key" => key.to_string()).increment(1);
 
-        if let Some(mut versioned_value) = value {
-            let value = StorageCodec::decode::<T, _>(&mut versioned_value.value)
-                .map_err(|err| ReadError::Codec(err.into()))?;
+        let value = self.inner.get(key.clone()).await?;
 
-            assert_eq!(
-                versioned_value.version,
-                value.version(),
-                "versions must align"
-            );
+        let result = value
+            .map(|mut versioned_value| {
+                let value = StorageCodec::decode::<T, _>(&mut versioned_value.value)
+                    .map_err(|err| ReadError::Codec(err.into()))?;
 
-            Ok(Some(value))
-        } else {
-            Ok(None)
-        }
+                assert_eq!(
+                    versioned_value.version,
+                    value.version(),
+                    "versions must align"
+                );
+
+                Ok(value)
+            })
+            .transpose();
+
+        histogram!(METADATA_CLIENT_GET_DURATION, "key" => key.to_string())
+            .record(start_time.elapsed());
+
+        result
     }
 
     /// Gets the current version for the given key. If key-value pair is not present, then return
     /// [`None`].
     pub async fn get_version(&self, key: ByteString) -> Result<Option<Version>, ReadError> {
-        self.inner.get_version(key).await
+        let start_time = Instant::now();
+        counter!(METADATA_CLIENT_GET_VERSION_TOTAL, "key" => key.to_string()).increment(1);
+        let result = self.inner.get_version(key.clone()).await;
+
+        histogram!(METADATA_CLIENT_GET_VERSION_DURATION, "key" => key.to_string())
+            .record(start_time.elapsed());
```

Comment:
That's a lot of histograms (cardinality is a problem here as well)

Reply:
Yeah, Indeed. How do you feel about the suggested solution above. Another solution of course it to drop the key label completely